src/Microsoft.AspNet.Http.Extensions/Internal/HeaderSegmentCollection.cs

```diff
@@ -16,7 +19,7 @@ public HeaderSegmentCollection(StringValues headers)
 
         public bool Equals(HeaderSegmentCollection other)
         {
-            return Equals(_headers, other._headers);
+            return StringValues.Equals(_headers, other._headers);
         }
```

Comment:
Currently both versions box equally; however change https://github.com/aspnet/Common/pull/55 is to resolve this and for a non-boxing path.